Peeking vol 1

DominikRafacz opened the pull request:
The use of `str` does not always allow the data to be summarized in a meaningful way (e.g. in the case of lists or data frames -- the first line of the structure summary carries almost no information. In this PR I introduce options that allow the user to choose the summary function and the size of this summary

DominikRafacz committed:
Merge pull request #1 from lewinfox/master

update

DominikRafacz committed:
Merge pull request #2 from lewinfox/master

merge

DominikRafacz committed:
remove already done todo

DominikRafacz committed:
add peeking

DominikRafacz committed:
max lines is now about length of summary

DominikRafacz committed:
update readme

DominikRafacz committed:
fix docs

DominikRafacz committed:
add autopeek function, set it as deafult peeking function

DominikRafacz committed:
missing imports

DominikRafacz committed:
rename x to object for consistency with str, docs fix

DominikRafacz committed:
update README

DominikRafacz commented:
I've added custom function, `ic_autopeek`, which is an improved version of `utils::str`. I'd really love to improve it even more, but -- honestly -- code of `str` is among the worst that I've ever seen. Those possible improvements may come in the future

ErdaradunGaztea committed:
add checkmate to use in tests

ErdaradunGaztea committed:
Merge branch 'feature/peeking' of https://github.com/ErdaradunGaztea/icecream into feature/peeking

ErdaradunGaztea committed:
create a few tests for ic_autopeek

ErdaradunGaztea committed:
fix truncated param description

ErdaradunGaztea committed:
add ic_autopeek tests

ErdaradunGaztea committed:
add ic_autopeek truncation tests

ErdaradunGaztea committed:
simplified test logic

ErdaradunGaztea committed:
fix case of the first element not fitting in supposed space

ErdaradunGaztea committed:
add test to cover fixed case

lewinfox commented:
Hey Dominik, thanks for this. I've poked around a little bit and it looks good!

Haven't read the code properly yet, but I'm making a start today. Might take me a little while so the comments will be rather piecemeal, apologies.

lewinfox commented on R/autopeek.R:
```diff
@@ -0,0 +1,68 @@
+#' @importFrom purrr map2_chr map_chr map_int detect_index
+#' @importFrom pillar obj_sum
+ic_autopeek_list_or_data.frame <- function(object,
+                                           max_summary_length = 70,
```
Replacing `max_summary_length = 70` with `max_summary_length = getOption("width")` means we will query the current terminal width each time. I've played around quickly and it seems to work pretty well.

Following on from that I've been using `ic(iris)` as an example. It shrinks the line length until only the first column is displayed, but perhaps you could modify the code for `ic_autopeek_list_or_data.frame()` so the minimum it would show is `... data.frame [<nrow> x <ncol>]`, i.e. not display any column summaries? That would make it look good on even very slim terminals.

I.e. at present the least info that will be displayed is

``` r
ic(iris)
#> ℹ ic| `iris`: data.frame [150 x 5]: $'Sepal.Length': dbl [150], ...
```

So if you have a _very_ narrow terminal (this example is 41 chars wide) you get wrapping:

```
                                         | <- terminal window edge
> ic(iris)                               |                               
ℹ ic| `iris`: data.frame [150 x 5]: $'Sep|
al.Length': dbl [150], ...               |
```

Whereas I think it would look nicer in these circumstances to lose the column summaries altogether:

```
                                         | <- terminal window edge
> ic(iris)                               |                               
ℹ ic| `iris`: data.frame [150 x 5]:      |
```

What do you think?

ErdaradunGaztea replied:
Hey Lewin, I'll answer this one on behalf of Dominik. I made a pull request to Dominik's branch where I did exactly what you described here and covered this with some tests on lists, so it should be solved once he finishes the breakfast and gets back to the computer.

DominikRafacz replied:
Hi, @ErdaradunGaztea, nice to see you there ;) I've reviewed your changes and merged them into my branch.

ErdaradunGaztea committed:
use glue vectorization and specialized functions

ErdaradunGaztea committed:
glue vectorization doesn't work when summary is 0-length vector

ErdaradunGaztea committed:
move checkmate to Suggests

DominikRafacz committed:
Merge pull request #3 from ErdaradunGaztea/feature/peeking

Some tests for autopeeking

lewinfox commented:
OK, I've reviewed everything. Really cool additions, thank you both :smile: 

My thoughts:

### :exclamation: Important functional changes

I think these are essential before we release.

- [ ] We need to `@export` S3 methods `ic_autopeek.default()`, `ic_autopeek.list()`, `ic_autopeek.data.frame()`, `ic_autopeek_header.default()`, `ic_autopeek_header.list()` and `ic_autopeek_header.data.frame()`.
- [ ] Unexport `ic_autopeek()` - users won't call this directly (correct me if I'm wrong) so it shouldn't be exported.

### :memo: Suggested code changes

I think these are non-essential but would improve the code. If you disagree with any of these let's discuss.

- [ ] Document `ic_autopeek_header()`. Doesn't have to be too detailed, but we should at least write a line or two explaining why it exists. As with `ic_autopeek()` this shouldn't be exported.
- [ ] This is a nit-pick but I would like to rename `ic_autopeek_list_or_data_frame()` to `ic_autopeek.list()` and then set `ic_autopeek.data.frame <- ic_autopeek.list`. I think this is a cleaner naming approach given that data frames are (sort of) a descendant of lists.
- [ ] In `ic_autopeek_list_or_data_frame()` (or renamed equivalent) replace `max_summary_length = 70` with `max_summary_length = getOption("width")` as mentioned in my comment above. I think this together with @ErdaradunGaztea 's changes will make for great printing behaviour.
- [ ] Add the documentation of new package options to `R/icecream-package.R` Can just copy-paste from the README, but it means that all docs are available even if you're not looking at GitHub.

### :nail_care:  Style suggestions

These are non-essential and a matter of personal preference :wink: 

- [ ] Run `usethis::use_tidy_style()` on the package.
- [ ] Replace `@importFrom pkg fun` with `pkg::fun()` in code. Although this makes the code visually messier I like the fact that it's easy to see where we are calling other packages.
- [ ] Wrap line lengths in `README.Rmd` to 100 chars.

What do you both think?

P.S. The PR is getting a little large so I'm going to set up a new release branch to house these changes.

lewinfox commented:
@DominikRafacz I've created a new feature branch at this end, might be worth rebasing your branch to make sure we're on track.

DominikRafacz committed:
changes in docs, reorganize data.frames and lists

DominikRafacz committed:
options docs

DominikRafacz committed:
rebuild docs

DominikRafacz committed:
restylize code

DominikRafacz committed:
wrap README to 100 chars

DominikRafacz commented:
Okay, so I've made some updates. I'm referring to your suggestions in order with some commentary here and there:

* [x] We need to `@export` S3 methods `ic_autopeek.default()`, `ic_autopeek.list()`, `ic_autopeek.data.frame()`, `ic_autopeek_header.default()`, `ic_autopeek_header.list()` and `ic_autopeek_header.data.frame()`.
* [x]  Unexport `ic_autopeek()` - users won't call this directly (correct me if I'm wrong) so it shouldn't be exported.

Well, in case of this point, I'm not sure. We're using this function only internally, but I can imagine the situation that somebody else would like to use improved `str` as well. The question is: shall it be the functionality of this package or shall there be another package for this... ? But for now maybe it's better to unexport this, as you've suggested.

* [x] Document ic_autopeek_header(). Doesn't have to be too detailed, but we should at least write a line or two explaining why it exists. As with ic_autopeek() this shouldn't be exported.
* [x]  This is a nit-pick but I would like to rename `ic_autopeek_list_or_data_frame()` to `ic_autopeek.list()` and then set `ic_autopeek.data.frame <- ic_autopeek.list`. I think this is a cleaner naming approach given that data frames are (sort of) a descendant of lists.

The reason I've named it this way is the fact that base R do not treat `data.frame` as `list` when dispatching methods (which is, in my opinion, a terrible mistake) and, in consequence, they are equal on syntactic level. This way of asigning the function to the names suggests in a way that lists are more important, more primal -- which is true on semiotic level, but not on syntactic level. However, these are just my thoughts and probably you're right. Your solution is simpler ;) 

* [ ]  In `ic_autopeek_list_or_data_frame()` (or renamed equivalent) replace `max_summary_length = 70` with `max_summary_length = getOption("width")` as mentioned in my comment above. I think this together with @ErdaradunGaztea 's changes will make for great printing behaviour.

This one is problematic. I agree that the `max_summary_length` should be dependent on `getOption("width")`, but it should also take into account the width of the prefix and expression, which are currently ommited. I can play with it and pass it as another argument.

* [x]  Add the documentation of new package options to `R/icecream-package.R` Can just copy-paste from the README, but it means that all docs are available even if you're not looking at GitHub.

 * [x]  Run `usethis::use_tidy_style()` on the package.
 * [ ]  Replace `@importFrom pkg fun` with `pkg::fun()` in code. Although this makes the code visually messier I like the fact that it's easy to see where we are calling other packages.

Well, here I will stick to my preference, but again -- a matter of taste ;) I will not change it now, but you are the maintainer, so feel free to do so :)

 * [x]  Wrap line lengths in `README.Rmd` to 100 chars.

DominikRafacz committed:
cat the result of peeking instead of returning it directly

DominikRafacz committed:
update tests to match autopeek interface

DominikRafacz committed:
reknit README

DominikRafacz commented:
@ErdaradunGaztea, replacement of `glue` with `paste0` in 730d962 was not perfect and I only noticed it after reknitting README.

Capturing output of `glue` in fact captures result of calling `cat` in `print.glue`. Capturing output of `paste0` captures result of calling `print.character`. As a result, in addition to the actual string, the index and qutotaion marks are printed. In order to fix this and keep consistency, I've added calling of `cat` inside `ic_autopeek` and then returning the string invisibly. In this way the behaviour is compliant with `glue` behaviour. 

Tests still worked fine after this change, but as the output of `cat` has been uncaught, I've replaced `expect_string` with `expect_output` to avoid printing the output in the console while testing.

lewinfox commented on R/autopeek.R:
```diff
@@ -0,0 +1,93 @@
+ic_autopeek <- function(object, ...) UseMethod("ic_autopeek")
+
+#' @export
+ic_autopeek.default <- str
```
Roxygen has exported this as a regular function rather than recognising it as an S3 method (see `NAMESPACE`). It looks like this is because Roxygen doesn't parse it correctly. We can remedy this by writing it so it looks more like a regular function:

``` r
#' @export
ic_autopeek.default <- function(object, ...) {
  str(object)
  # Should this be `str(object, ...)`?
}
```

or by tagging it by hand:

``` r
#' @method ic_autopeek default
#' @export
ic_autopeek.default <- str
```

DominikRafacz replied:
Thanks for pointing this out, fixing the issue in f30dc09

lewinfox commented:
> The reason I've named it this way is the fact that base R do not treat data.frame as list when dispatching methods (which is, in my opinion, a terrible mistake) and, in consequence, they are equal on syntactic level. This way of asigning the function to the names suggests in a way that lists are more important, more primal -- which is true on semiotic level, but not on syntactic level. However, these are just my thoughts and probably you're right. Your solution is simpler ;)

Yeah I was uneasy writing that comment to be honest as it's a grey area - whatever you think is best.

> Well, here I will stick to my preference, but again -- a matter of taste ;) I will not change it now, but you are the maintainer, so feel free to do so :)

Entirely reasonable :smile: If I find myself with time on my hands in future I may pick that up. 

I've added one more comment above about `ic_autopeek.default()` and then I think we can merge the branch. I think  this plus #6 would be a nice 0.2 release, would you agree?

DominikRafacz committed:
export ic_autopeek.default

DominikRafacz commented:
> I've added one more comment above about `ic_autopeek.default()` and then I think we can merge the branch. I think this plus #6 would be a nice 0.2 release, would you agree?

@lewinfox, Totally agree! This set of changes is sufficiently extensive to merit its own release :) But in the forthcoming days I won't have enough time for implementing #6, so I'll be really grateful if you take matters into your own hands.
